### ISSUE TITLE
litcoffee source does not make for easy debugging or contributing.

### ISSUE THREAD

**Issue by AdrianRossouw** [nickfargo/state#14]
While I understand, and really appreciate the wonderfully complete documentation that this format affords, it caused me a significant amount of problems trying to debug an issues that I was having.

The module is just delivered to the environment as pre-compiled javascript source, with no source maps and no real easy way to generate them or use them. Once you get into stack traces and step debugging of some of the more interesting functions you end up being faced with half a dozen _refN variables that all look the same..

On a personal note it was also a constant battle to decide wether or not to spend the estimated hour or so to train vim to properly understand that these markdown files (.md extension and all) were supposed to be read as coffeescript and to do the highlighting and indentation properly. Even after all that there was going to be nearly no relation towards what would actually be executed.

I think that there has to be some way to map the litcoffee to an intermediary coffeescript format, that can have proper source maps built for them. And at the very least please use an appropriate file extension so that syntax highlighting and other build tools can understand the source too.

The website and documentation is truly wonderful, and a very large reason why I ended up picking this library, but I found the cost to be really high to dive into the code itself.

**Comment by nickfargo** [comment]
I hear you … along with the obvious advantages, one of the more obvious tradeoffs of moving from plain JS to literate Coffee is the apparent additional barrier-to-entry for contributors.

At the same time, most of the difficulty, as I see it now, is actually related to the project structure and build strategy. Presently the project structure follows in the Node style, where each “class” is in its own file and therefore a separate module; the entire `/src` directory is then compiled by `coffee` (via the `bin/compile` script) into JS under `/lib`, and this directory is finally handed off to browserify. So, in terms of source maps, the current problem, of course, is that neither step is aware of the other.

What I still need to learn is whether it’s feasible to create, either through browserify or some other mature tool, or manually, higher-order source maps that can navigate multiple compile steps (i.e. literate coffee -> compiled js -> uglified js), resulting in a single **.map** that points directly back to a neatly structured tree of coffee files.

Absent that, in terms of workflow, to this point I’ve just been working the day-to-day in Node, trying to stay covered (mostly) on writing tests, at least for any new place I go a-hacking. In my experience so far, the errors I encounter have been — let’s say _informative enough_ — and in practice, really not too hard to stack-trace right back to the coffee source.

---

As to the **.coffee.md** pseudo-extension … I am partial, though not wedded, to the idea of the files doubling as “dumb” markdown, for agents that can’t (and shouldn’t necessarily be expected to) read them as executable code. As of Coffee 1.6.3, unless I’m much mistaken this should now just be a matter of getting the tooling to play nice …

Personally I’ve had good success in Sublime with **.coffee.md** using [the tmbundle](https://github.com/jashkenas/coffee-script-tmbundle/blob/master/Syntaxes/CoffeeScript%20%28Literate%29.tmLanguage#L11) from @jashkenas, which does fairly well on its own, though I’ve also added a few hacks of my own, and some corresponding color scheming, to make it a little bit smarter/prettier — I guess that’s something that I really should release as a separate repo/fork sometime.

![Literate CoffeeScript in SublimeText](https://f.cloud.github.com/assets/578266/1595519/94df9cf8-52eb-11e3-9daf-92cb8b9fcaa1.png)

I’m an unenlightened vim non-user, however — is it the case then that the existing tools such as [vim-literate-coffeescript](https://github.com/mintplant/vim-literate-coffeescript) do not recognize **.coffee.md** as literate coffee?

![Literate CoffeeScript in vim](https://github.com/mintplant/vim-literate-coffeescript/raw/master/test/test.png)

**Comment by AdrianRossouw** [comment]
The general rule of thumb for external libraries is that you are most likely to be ignored by a fairly large subset of developers if you are written in coffeescript. I've seen a lot of libraries get ported back to JS to get more users.

That said, because of the nature of the problem domain, I agree that litcoffee is actually a pretty good way to handle what is a fairly enormous documentation task. 

I was in no way suggesting you should change how the software is developed. I was suggesting that the build system be improved to provide a more accessible experience to people new to the system. 

The good news is that browserify can definitely handle all that. I have spent a lot of time with browserify ([i even have some slides to prove it](http://vertice.github.io/browserify-slides/)), so i am fairly certain of this. 

I'm personally rather partial towards tying it together using grunt, because some of the heavy lifting relating to dependencies is much much easier that way, but seeing as there are hardly any dependencies a simple shell script should suffice.

My concern with tool support is more about situations where you might need a grunt plugin or similar that just doesn't recognize the .coffee.md format as actual code. It does seem like the [.coffee.md is set up right here](https://github.com/mintplant/vim-literate-coffeescript/blob/master/ftdetect/vim-literate-coffeescript.vim#L6). In my defense though, it was pretty late to be fighting my environment too. I also reserve judgement on how pleasant litcoffee will be in node-inspector.

**Comment by nickfargo** [comment]
> The general rule of thumb for external libraries is that you are most likely to be ignored by a fairly large subset of developers if you are written in coffeescript.

Yeah. And, to an extent, I’m ok with that. Porting back to JS has been put down as a real possibility since day one of porting over to (literate) Coffee, though if that were to happen, as it stands now it would be for technical reasons — e.g., to pursue ES6 features, which could then be Traceur’d into everywhere-JS during a compile step, or something — rather than a play to gain user share.

Grunt usage is small but will likely increase — right now there’s just a task for pumping all the literate source through a fork of docco to spit out the prettified HTML under [/source/](http://statejs.org/source/).

I’ll be very interested to see whether this all can eventually be neatly source-mapped and rigged up for stepping in node-inspector, while also keeping the file/directory structure intact — i.e. not having to cat all the sources into one monolithic **.coffee** file — yech

Whatever the build process is, and however it evolves, perhaps that needs to be documented just as rigorously as the project itself.

Thanks for the ideas … much to mull over …

**Comment by AdrianRossouw** [comment]
I actually spent an evening porting a project of mine to ES6 just to see what it was like. The toolchain is still pretty immature, and I think the fat arrow functions will cause sheer havoc with state.js. They are always bound to the context function in which they are defined. state.bind and friends would no longer work.

I played around with browserify, and this is how far I got :

### client side static build

```
browserify --standalone state --debug --transform coffeeify --extension .coffee.md src/index.coffee.md > lib/browser.js
```

This will generate a standalone package in UMD format that will work straight in the browser, through require.js (blech) and as a commonjs module. This package could also be published and used straight from bower, or whatever client-side packaging tool you use.

You can then do this to your package.json , which will load up a different file when browserifying it, and it will just silently add it to the browser.

```
    "main": "lib/index",
+  "browser": "lib/browser",
    "directories": {
```

### dynamic browserify build

On the opposite side ... if you add coffeeify to the project, and add this to your package.json

```
   "main": "src/index",
+ "browserify": { "transforms": "coffeeify" },
   "directories": {
```

theoretically browserify will know to pre-compile the coffeescript through the coffeeify transform,
as the code is being bundled for the client .. but the issue here is that the .coffee.md extension
still needs to be added, and that doesn't seem to be passed along to the lower end things yet.

(this is where i why i was worried about the non-standard extension btw..)

What this means, is that to trigger the coffeeify transform, you would have to require all the
files with absolute paths - ie: require('./state.coffee.md') That's not ideal, but there is another
option that can work too, but can be a little finnicky :

The browser field above has a different form, that allows you to map packages to source files,
so you should be able to map './state': './src/state.coffee.md'.

### and the source maps

While it's definitely possible to generate the source maps this way, i'm not sure how deeply nested
the debug argument to browserify goes, ie: if you have state in your package.json, will it end up passing
the source map down to the client, for the dependencies.

From my limited testing, some of the modules do this .. others don't. I will have to go look into it more,
but i have half a dozen issues / pull requests archived on the subject.

### on the node.js side of things

since we would be generating valid maps in various different ways, the accepted way to enable
source maps on the server side would be using : https://github.com/evanw/node-source-map-support

untested. should work fine.

### next steps

I started a separate little project to experiment with, and document the best practices I find for some of the trickier packaging problems I come across. It's much easier to reason about these things when you are in a cleanroom environment without your application code muddying things up.

So I'll wire up the state module to my [browserify testing harness](https://github.com/Vertice/browserify-harness), when I get the chance, and see if I can come up with the best approach that hit's all the requirements.

**Comment by nickfargo** [comment]
Brilliant!

Thanks for the recon … I’ll get this on the to-do list and start experimenting with some of the build options.
